### ISSUE TITLE
codeql says current master is affected by code injection but shows past commits

### ISSUE THREAD

**Issue by darkbasic** [github/codeql#14935]
**Description of the false positive**

In Accounts.js impersonation is not a feature that could be easily exploited because it must be enabled via a custom `options.impersonationAuthorize` function, but this is not the point. The point is that there is a potential MongoDB Injection. I don't see how that could happen considering the `.toLowerCase()` usage but that's still not the point. From codeql's perspective user input must be sanitized and it's a good practice to do regardless, so let's analize what it has to say. Clicking on `Show paths` shows where we get the user input and goes through all the code until the code injection happens. Let's see what we have on current master:

**Code samples or links to source code**

User input starts from here:

https://github.com/accounts-js/accounts/blob/master/packages/rest-express/src/endpoints/impersonate.ts#L18

Code injection happens here:

https://github.com/accounts-js/accounts/blob/master/packages/database-mongo-password/src/mongo-password.ts#L204

I'm no MongoDB injection expert, but usually injections are done passing an object like `{ "$gt": "" }`. If you pass the string `'{ "$gt": "" }'` you won't be able to achieve anything, so the `body('impersonated.username').optional().isString()` express-validator middleware should successfully sanitize the user input here.

From what I can see no alert should be shown, so why do we have one in the first place? express-validator has been introduced recently so the alert makes sense but somehow it hasn't been closed when [this commit](https://github.com/accounts-js/accounts/commit/a0155f50aeb0ef13db3c215227713b7a12114a27) landed.

![image](https://github.com/github/codeql/assets/1047358/bfb2ac65-8429-4423-bd09-a61cb62ff0bb)

The first thing I notice is that under `Affected branches` we have two configurations.
[.github/workflows/codeql-analysis.yml](https://github.com/accounts-js/accounts/blob/master/.github/workflows/codeql-analysis.yml) has been recently updated to use `javascript-typescript` instead of `javascript`.
So I guess that I should click on the trash icon near `.github/workflows/codeql-analysis.yml:analyze/language:javascript` because somehow it's still picking the old workflow.
The second item is called `/language:javascript-typescript`: shouldn't it be called `.github/workflows/codeql-analysis.yml:analyze/language:javascript-typescript` instead? It appears to have been triggered last week while it should have been triggered by yesterday's push to master as well: https://github.com/accounts-js/accounts/actions/runs/7008342986

![image](https://github.com/github/codeql/assets/1047358/6aa9ef9b-55c1-4ebf-b507-3c8528e374df)

If I click on `Show paths` it shows an old version of the code previous to the [express-validator commit](https://github.com/accounts-js/accounts/commit/a0155f50aeb0ef13db3c215227713b7a12114a27).

![image](https://github.com/github/codeql/assets/1047358/0effdf54-53c6-4b84-b87a-34cb861b7f0f)

Looking at the `fixed in branch master` I notice that it shows a commit that does not belong to any branch in the repository. This is my bad because I'm developing alone lately and I force pushed to master a couple of times to get things done quicker. I'm wondering if this has anything to do with the fact that it didn't pick up any further scans that should have been triggered by more recent commits.

**URL to the alert on GitHub code scanning (optional)**

https://github.com/accounts-js/accounts/security/code-scanning/12

**Comment by jketema** [comment]
Hi @darkbasic,

It indeed seems that the alert was not closed for some reason. Have you tried re-running your CodeQL workflow on `master`?

**Comment by darkbasic** [comment]
> Have you tried re-running your CodeQL workflow on master?

What do you mean by re-running? You mean triggering it with a new commit on master? I've done so many times already.

![image](https://github.com/github/codeql/assets/1047358/7c06cf34-ecd5-4ded-8643-c472001a609b)

Under code security is says that last scan has been made 17 hours ago and that appears coherent with latest action run:

![image](https://github.com/github/codeql/assets/1047358/3ab322b7-bc9c-4fb1-a569-d49fea2e59fe)

**Comment by cannist** [comment]
It looks like the configuration was changed with commit https://github.com/accounts-js/accounts/commit/17410d62ba85db28c348d614cfc912d91d4b7192  and an explicit category name configured. This led to a state where code scanning knows about two configurations for analyzing the code: an older one (implicitly) named `.github/workflows/codeql-analysis.yml:analyze/language:javascript` that is not run anymore and a newer one explicitly named `/language:javascript-typescript`. The name for the latter is defined in the workflow file [here](https://github.com/accounts-js/accounts/commit/17410d62ba85db28c348d614cfc912d91d4b7192#diff-63bd641104d10e25f141d518a16b22a151d125e12701df2f9e79734b23b90188R81).

The alert has been detected in both configurations but the fix has only been observed by the newer one. Indeed, the way to solve this is to get rid of the old configuration by clicking the trash icon for the old config. This is the right thing to do since you have essentially replaced the old configuration with the new one and the old one will not be run anymore.

![image](https://github.com/github/codeql/assets/1047358/bfb2ac65-8429-4423-bd09-a61cb62ff0bb)

> The second item is called /language:javascript-typescript: shouldn't it be called .github/workflows/codeql-analysis.yml:analyze/language:javascript-typescript instead? 

The name is explicitly set in the workflow file via the `category` parameter.

> It appears to have been triggered last week while it should have been triggered by yesterday's push to master as well:

In the configuration `/language:javascript-typescript` the alert is considered fixed which is indicated by the purple shield-with-a-checkmark icon. For the fixed alert `last week` refers to the time when the fix has been observed, i.e. the first time the alert was not found anymore. It does not reflect further analyses that might have happened after that.


That we're showing code from an older commit is a consequence of the fact that newer commits that have been analyzed do not have the alert anymore, so we go to the old commit that had it to be able to show it.

---

To summarize: everything should get fixed by deleting the stale configuration `.github/workflows/codeql-analysis.yml:analyze/language:javascript` via the trash icon. Once it is gone the alert will be considered `fixed` because it is fixed in the only remaining configuration.

**Comment by darkbasic** [comment]
Thanks, that was indeed the case.
